src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,993 @@
+        return (sparse.sizeofNode(node) < dense.sizeofNode(node)) ? sparse : dense;
+    }
+
+    /**
+     * Returns the size needed to serialize this node.
+     */
+    abstract public int sizeofNode(SerializationNode<?> node);
+
+    /**
+     * Serializes the node. All transition target positions must already have been defined. {@code payloadBits} must
+     * be four bits.
+     */
+    abstract public void serialize(DataOutputPlus out, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException;
+
+    // Implementations
+
+    final int bytesPerPointer;
+    static final int FRACTIONAL_BYTES = 0;
+
+    TrieNode(int ordinal, int bytesPerPointer)
+    {
+        this.ordinal = ordinal;
+        this.bytesPerPointer = bytesPerPointer;
+    }
+
+    final int ordinal;
+
+    static final TrieNode PAYLOAD_ONLY = new PayloadOnly();
+
+    static private class PayloadOnly extends TrieNode
+    {
+        // byte flags
+        // var payload
+        PayloadOnly()
+        {
+            super(0, FRACTIONAL_BYTES);
+        }
+
+        @Override
+        public int payloadPosition(ByteBuffer src, int position)
+        {
+            return position + 1;
+        }
+
+        @Override
+        public int search(ByteBuffer src, int position, int transitionByte)
+        {
+            return -1;
+        }
+
+        @Override
+        public long transitionDelta(ByteBuffer src, int position, int childIndex)
+        {
+            return 0;
+        }
+
+        @Override
+        public long transition(ByteBuffer src, int position, long positionLong, int childIndex)
+        {
+            return -1;
+        }
+
+        @Override
+        public long lastTransition(ByteBuffer src, int position, long positionLong)
+        {
+            return -1;
+        }
+
+        @Override
+        public long greaterTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            return defaultValue;
+        }
+
+        @Override
+        public long lesserTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            return defaultValue;
+        }
+
+        @Override
+        public int transitionByte(ByteBuffer src, int position, int childIndex)
+        {
+            return Integer.MAX_VALUE;
+        }
+
+        @Override
+        public int transitionRange(ByteBuffer src, int position)
+        {
+            return 0;
+        }
+
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            return 1;
+        }
+
+        @Override
+        public void serialize(DataOutputPlus dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+        }
+    }
+
+    static final TrieNode SINGLE_8 = new Single(2, 1);
+    static final TrieNode SINGLE_16 = new Single(4, 2);
```

Comment:
Aside: I think we moved these into `Single` itself in one of the SAI patches to avoid potential class loading deadlocks.

Reply:
See https://github.com/apache/cassandra/pull/1466/commits/b45f47d7097b2c9fbf6abe1c10ae60543b0068cb

Reply:
As the subclasses are now private (see https://github.com/apache/cassandra/pull/2267/commits/8d5de21b48624301875a31d4e35bbc3bf8c1a68f), there is no possibility for concurrent load of class and subclass and hence no deadlock.

The construction was also changed somewhat to make the ordinal value final.

Reply:
hmmm, IDEA's static analysis tooling still complains about it...perhaps just a bug there

Reply:
I gave up and moved all the static instances to an embedded `Types` class.